--- mesa-13.0.1/src/compiler/glsl/cache.c.orig	2016-11-21 16:37:20.000000000 +0100
+++ mesa-13.0.1/src/compiler/glsl/cache.c	2016-11-21 16:47:16.000000000 +0100
@@ -222,7 +222,23 @@
    if (path == NULL)
       goto fail;
 
+#ifdef O_CLOEXEC
    fd = open(path, O_RDWR | O_CREAT | O_CLOEXEC, 0644);
+   if (fd == -1 && errno == EINVAL)
+#endif
+   {
+      fd = open(path, O_RDWR | O_CREAT, 0644);
+      if (fd == -1)
+          goto fail;
+
+      /* Snow Leopard doesn't have O_CLOEXEC */
+      int cloexec = fcntl(fd, F_SETFD, FD_CLOEXEC);
+      if (cloexec == -1) {
+         close(fd);
+         goto fail;
+      }
+   }
+#endif
    if (fd == -1)
       goto fail;
 
